language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala

```diff
@@ -109,78 +55,81 @@ object CodeGen {
   @throws[PackageInterfaceException](
     cause = "either decoding a package from a file or extracting" +
       " the package interface failed")
-  def generateCode(
-      sdamlFile: File,
-      otherDalfInputs: Seq[URL],
+  def generateCode(files: List[File], packageName: String, outputDir: File, mode: Mode): Unit = {
+    generateCodeSafe(files, packageName, outputDir, mode)
+      .fold(es => throw PackageInterfaceException(formatErrors(es)), identity)
+  }
+
+  private def formatErrors(es: NonEmptyList[String]): String =
+    es.toList.mkString("\n")
+
+  def generateCodeSafe(
+      files: List[File],
       packageName: String,
       outputDir: File,
-      mode: Mode): Unit = {
-    val errorOrRun = for {
-      interface <- decodePackageFromFile(sdamlFile, mode)
-      dependencies <- decodePackagesFromURLs(otherDalfInputs, mode)
-      combined = mode.combineInterfaces(interface, dependencies)
-    } yield packageInterfaceToScalaCode(mode.Dialect(packageName, combined, outputDir))
-
-    errorOrRun fold (e => throw PackageInterfaceException(e), identity)
-  }
+      mode: Mode): ValidationNel[String, Unit] = {
 
-  private def decodePackageFromFile(
-      sdamlFile: File,
-      mode: Mode): String \/ mode.InterfaceElement = {
-    val is = Files.newInputStream(sdamlFile.toPath)
-    println(s"Decoding ${sdamlFile.toPath}")
-    try decodePackageFrom(is, mode)
-    finally is.close() // close is in case of fatal throwables
-  }
+    val reader = UniversalArchiveReader()
 
-  private def decodePackagesFromURLs(
-      urls: Seq[URL],
-      mode: Mode): String \/ Seq[mode.InterfaceElement] =
-    urls
-      .map { url =>
-        val is = url.openStream()
-        try decodePackageFrom(is, mode)
-        finally is.close()
-      }
-      .toList
-      .sequenceU
+    val errorsOrUnits: ValidationNel[String, List[Unit]] = files.traverseU { f =>
```

Comment:
I believe this traversal will create separate runs for each sub-`EnvironmentInterface`, producing multiple EventDecoder and package ID files for multi-argument situations, probably overwriting as it goes. If so, we need to handle the whole argument set with a single `decodeInterface` to produce a single `EnvironmentInterface` to do the whole codegen run on.

Reply:
this is something worth of a test case... I will create a second daml file for the Integration Test.